docs/messaging/messaging-spans.md

```diff
@@ -195,21 +197,64 @@ Examples:
 * `AuthenticationRequest-Conversations process`
 * `(anonymous) publish` (`(anonymous)` being a stable identifier for an unnamed destination)
 
-### Span kind
-
-A producer of a message should set the span kind to `PRODUCER` unless it synchronously waits for a response: then it should use `CLIENT`.
-The processor of the message should set the kind to `CONSUMER`, unless it always sends back a reply that is directed to the producer of the message
-(as opposed to e.g., a queue on which the producer happens to listen): then it should use `SERVER`.
-
 ### Operation names
 
 The following operations related to messages are defined for these semantic conventions:
 
 | Operation name | Description |
 | -------------- | ----------- |
-| `publish`      | A message is sent to a destination by a message producer/client.       |
-| `receive`      | A message is received from a destination by a message consumer/server. |
-| `process`      | A message that was previously received from a destination is processed by a message consumer/server. |
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER`, |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+### Trace structure
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
```

Comment:
"Create" spans MAY be created." seems redundant

Reply:
Creating a "Publish" span is a strong requirement, it needs to present in all producer scenarios. We formulated it as a SHOULD.

Creating "Create" spans is a weaker requirement. It shouldn't be created for single-message scenarios and for scenarios where a creation context is provided. That's why we settled on a MAY, if I remember correctly.

Here's the [relevant part from the OTEP](https://github.com/open-telemetry/oteps/blob/main/text/trace/0220-messaging-semantic-conventions-span-structure.md#producer):

> There are four different scenarios for injecting a creation context into a message:
>
> 1. A user provides custom creation contexts for the messages that are published, which don't refer to any spans described in this document. This provides flexibility to users to model custom scenarios. In this case, no other additional spans besides the "Publish" span should be created. The "Publish" span should link to the provided creation contexts.
> 2. If no custom creation context is provided for a message, it is recommended to create a "Create" span for every single message and inject its context into the message. "Create" spans can be created during the "Publish" operation as children of the "Publish" span.
> 3. As a variation of the scenario above, "Create" spans can be created independently of the "Publish" operation, e. g. in cases where messages are created before they are passed to a "Publish" operation. In this case, the "Publish" span should link to the "Create" spans.
> 4. For single-message scenarios or when large number of spans are a problem, the context of the "Publish" span can be injected into the message, thus acting as the creation context. In this case, no other spans besides the "Publish" span should be created.

Reply:
Discussed in today's SIG meeting: try to explain operations in more detail in the operation names table (in the description).